Neos.Neos/Classes/Controller/Service/AssetProxiesController.php

```diff
@@ -0,0 +1,173 @@
+use Neos\Neos\Controller\BackendUserTranslationTrait;
+use Neos\Neos\View\Service\AssetJsonView;
+
+/**
+ * Rudimentary REST service for assets
+ *
+ * @Flow\Scope("singleton")
+ */
+class AssetProxiesController extends ActionController
+{
+    use BackendUserTranslationTrait;
+
+    /**
+     * @Flow\Inject
+     * @var AssetRepository
+     */
+    protected $assetRepository;
+
+    /**
+     * @Flow\Inject
+     * @var AssetSourceService
+     */
+    protected $assetSourceService;
+
+    /**
+     * @Flow\Inject
+     * @var TagRepository
+     */
+    protected $tagRepository;
+
+    /**
+     * @Flow\InjectConfiguration(package="Neos.Media", path="asyncThumbnails")
+     * @var boolean
+     */
+    protected $asyncThumbnails;
+
+    /**
+     * @var array
+     */
+    protected $viewFormatToObjectNameMap = array(
+        'html' => TemplateView::class,
+        'json' => AssetJsonView::class
+    );
+
+    /**
+     * A list of IANA media types which are supported by this controller
+     *
+     * @var array
+     * @see http://www.iana.org/assignments/media-types/index.html
+     */
+    protected $supportedMediaTypes = array(
+        'text/html',
+        'application/json'
+    );
+
+    /**
+     * @param ViewInterface $view
+     * @return void
+     */
+    public function initializeView(ViewInterface $view)
+    {
+        $view->assign('asyncThumbnails', $this->asyncThumbnails);
+    }
+
+    /**
+     * Shows a list of assets
+     *
+     * @param string $searchTerm An optional search term used for filtering the list of assets
+     * @param string $assetSourceIdentifier If specified, results are only from the given asset source
+     * @param int $limit The maximum number of results shown in total
+     * @return void
+     */
+    public function indexAction(string $searchTerm = '', string $assetSourceIdentifier = '', int $limit = 10): void
+    {
+        $assetProxyQueryResultsIterator = new \MultipleIterator(\MultipleIterator::MIT_NEED_ANY);
+        foreach ($this->assetSourceService->getAssetSources() as $assetSource) {
+            if ($assetSourceIdentifier === '' || $assetSource->getIdentifier() === $assetSourceIdentifier) {
+                $assetProxyQueryResult = $assetSource->getAssetProxyRepository()->findBySearchTerm($searchTerm);
+                if ($assetProxyQueryResult->valid()) {
+                    $assetProxyQueryResultsIterator->attachIterator($assetProxyQueryResult);
+                }
+            }
+        }
+
+        $totalAddedResults = 0;
+        $assetProxiesByAssetSource = [];
+        foreach ($assetProxyQueryResultsIterator as $assetProxies) {
+            foreach ($assetProxies as $assetProxy) {
+                if ($assetProxy instanceof AssetProxyInterface) {
+                    $assetProxiesByAssetSource[$assetProxy->getAssetSource()->getIdentifier()][] = $assetProxy;
+                    $totalAddedResults ++;
+                }
+                if ($totalAddedResults === $limit) {
```

Comment:
nitpick: that change can be moved inside the if condition above as that's the only part that alters the counter
Or even safe a nesting level by doing
```php
foreach ($assetProxies as $assetProxy) {
    if (!$assetProxy instanceof AssetProxyInterface) {
        continue;
    }
    $assetProxiesByAssetSource[$assetProxy->getAssetSource()->getIdentifier()][] = $assetProxy;
    if (++$totalAddedResults === $limit) {
        break 2;
    }
}
```
but I know that you dislike `continue` :)